platform/security/certificate/security-crl-generator/src/main/java/org/codice/ddf/security/crl/generator/CrlGenerator.java

```diff
@@ -0,0 +1,315 @@
+import org.codice.ddf.system.alerts.SystemNotice;
+import org.osgi.service.event.Event;
+import org.osgi.service.event.EventAdmin;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The CrlGenerator downloads and saves a Certificate Revocation List from a URL specified by {@code
+ * crlLocationUrl}.
+ */
+public class CrlGenerator implements Runnable {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(CrlGenerator.class);
+
+  private static final PropertiesLoader PROPERTIES_LOADER = PropertiesLoader.getInstance();
+
+  private static final String HTTPS = "https://";
+
+  private static final int INITIAL_DELAY = 0;
+
+  private static final int SCHEDULER_INTERVAL = 30;
+
+  private static final int FILE_BACKED_STREAM_THRESHOLD = 1000000;
+
+  static final String CRL_PROPERTY_KEY = Merlin.OLD_PREFIX + Merlin.X509_CRL_FILE;
+
+  static final String DEM_CRL = "/localCrl.crl";
+
+  static final String PEM_CRL = "/localCrl.pem";
+
+  @VisibleForTesting
+  static String issuerEncryptionPropertiesLocation =
+      new AbsolutePathResolver("etc/ws-security/issuer/encryption.properties").getPath();
+
+  @VisibleForTesting
+  static String issuerSignaturePropertiesLocation =
+      new AbsolutePathResolver("etc/ws-security/issuer/signature.properties").getPath();
+
+  @VisibleForTesting
+  static String serverEncryptionPropertiesLocation =
+      new AbsolutePathResolver("etc/ws-security/server/encryption.properties").getPath();
+
+  @VisibleForTesting
+  static String serverSignaturePropertiesLocation =
+      new AbsolutePathResolver("etc/ws-security/server/signature.properties").getPath();
+
+  @VisibleForTesting
+  static String crlFileLocation = new AbsolutePathResolver("etc/keystores").getPath();
+
+  private final ClientFactoryFactory factory;
+
+  private final EventAdmin eventAdmin;
+
+  private final ScheduledExecutorService scheduler;
+
+  private String crlLocationUrl;
+
+  private boolean crlByUrlEnabled;
+
+  private ScheduledFuture<?> handle;
+
+  public CrlGenerator(ClientFactoryFactory factory, EventAdmin eventAdmin) {
+    this.factory = factory;
+    this.eventAdmin = eventAdmin;
+    this.scheduler =
+        Executors.newSingleThreadScheduledExecutor(
+            StandardThreadFactoryBuilder.newThreadFactory("crlThread"));
+  }
+
+  /** Pulls down the CRL file provided via URL and writes it locally. */
+  @Override
+  public synchronized void run() {
+
+    if (!crlByUrlEnabled || crlLocationUrl == null) {
+      return;
+    }
+
+    if (!crlLocationUrl.startsWith(HTTPS)) {
+      postErrorEvent("The provided URL was not an HTTPS URL.");
+      return;
+    }
+
+    SecureCxfClientFactory cxfClientFactory =
+        factory.getSecureCxfClientFactory(crlLocationUrl, WebClient.class);
+    WebClient client = cxfClientFactory.getWebClient();
+    Response response = client.get();
+    Object entity = response.getEntity();
+
+    if (entity instanceof InputStream) {
```

Comment:
✏️ This is a large block of nested conditions/try-catches. It probably should be broken up into meaningfully named methods to extract out the details and keep this method operating at a single, higher cognitive level.

Reply:
Maybe something like this would work - https://gist.github.com/coyotesqrl/5d1eff733a9bae746dea7fee7883f9fe

Reply:
I can do everything but the `readResponseContent`. I want to make sure I call `fileBackedOutputStream.reset()` at the end to delete the temp file tht is created (if it's created).